fixlib/fixlib/baseresources.py

```diff
@@ -1612,4 +1613,77 @@ def delete(self, graph: Any) -> bool:
         return False
 
 
+class PolicySourceKind(StrEnum):
+    Principal = "principal"  # e.g. IAM user, attached policy
+    Group = "group"  # policy comes from an IAM group
+    Resource = "resource"  # e.g. s3 bucket policy
+
+
+ResourceConstraint = str
+
+ConditionString = str
+
+
+@frozen
+class PolicySource:
+    kind: PolicySourceKind
+    uri: str
+
+
+class HasResourcePolicy(ABC):
+    # returns a list of all policies that affects the resource (inline, attached, etc.)
+    def resource_policy(self, builder: Any) -> List[Tuple[PolicySource, Json]]:
+        raise NotImplementedError
+
+
+@frozen
+class PermissionCondition:
+    # if nonempty and any evals to true, access is granted, otherwise implicitly denied
+    allow: Optional[Tuple[ConditionString, ...]] = None
+    # if nonempty and any is evals to false, access is implicitly denied
+    boundary: Optional[Tuple[ConditionString, ...]] = None
+    # if nonempty and any evals to true, access is explicitly denied
+    deny: Optional[Tuple[ConditionString, ...]] = None
+
+
+@frozen
+class PermissionScope:
+    source: PolicySource
+    constraints: Tuple[ResourceConstraint, ...]  # aka resource constraints
+    conditions: Optional[PermissionCondition] = None
+
+    def with_deny_conditions(self, deny_conditions: List[Json]) -> "PermissionScope":
+        c = self.conditions or PermissionCondition()
+        return evolve(self, conditions=evolve(c, deny=tuple([to_json_str(c) for c in deny_conditions])))
+
+    def with_boundary_conditions(self, boundary_conditions: List[Json]) -> "PermissionScope":
+        c = self.conditions or PermissionCondition()
+        return evolve(self, conditions=evolve(c, boundary=tuple([to_json_str(c) for c in boundary_conditions])))
+
+    def has_no_condititons(self) -> bool:
+        if self.conditions is None:
+            return True
+
+        if self.conditions.allow is None and self.conditions.boundary is None and self.conditions.deny is None:
+            return True
+
+        return False
+
+
+class PermissionLevel(StrEnum):
+    List = "list"
+    Read = "read"
+    Tagging = "tagging"
+    Write = "write"
+    PermissionManagement = "permission"
+    Unknown = "unknown"  # in case a resource is not in the levels database
```

Comment:
```suggestion
    list = "list"
    read = "read"
    tagging = "tagging"
    write = "write"
    permissionManagement = "permission"
    unknown = "unknown"  # in case a resource is not in the levels database
```